### DISCUSSION TITLE
Why does `quarto preview` always choose a new port?

### DISCUSSION THREAD

**Discussion by SamEdwardes** [quarto-dev/quarto-cli#2083] [Q&A] [answered]
When you run the command `quarto preview` the server will use a different port each time. I know that I can specify the port I want to use using `quarto preview --port xxxx`. However, I am curious why the choice was made to choose a different port each time? In my (anecdotal) experience, other frameworks like Dash and Streamlit always use the same port.

In my typical workflow, I will be editing my .qmd files with the preview server running. Sometimes the server may crash (for example I switch branches). Then I need to run `quarto preview` again. The slight annoyance is that I cannot refresh my browser window that had site up and running.

**Comment by cscheid** [comment, marked as answer]
We do this because we've found in the past that some operating systems sometimes hold ports longer than the process's lifetime, which means that quickly iterating through a number of `quarto preview` processes could lead to users having to manually choose different ports anyway.

(I find this behavior not optimal either, but the alternative is worse)

**Reply by SamEdwardes** [reply]
Thank you for explaining why!

**Reply by rgaiacs** [reply]
Can you be more specific about `some operating systems`? Is this been reported in only some systems?

**Reply by mcanouil** [reply]
To note, this is standard practice to expose random ports when not deploying any services to production.
For example, in VSCode, extensions uses port to display help page, debugger, etc.
Sometimes, after a crash, error in the code, long computation, the port is not properly freed because the process is still on leading to the whole thing being unavailable.
This does not happen when exposing random ports, because new free ports can be used.

**Reply by cbrnr** [reply]
Thanks for the explanation. I'd also prefer the default to be some constant port, but now at least I know your reasons for this choice!

**Comment by neeldhara** [comment]
In my `quarto.yml` file, I have:

```
project:
  type: website
  preview:
    port: 7777
```

and this fixes the port to 7777.

I had the exact same issue as the OP, and found this workaround in the docs somewhere :)

**Reply by mcanouil** [reply]
This is from <https://quarto.org/docs/reference/projects/options.html#preview> 😉

**Reply by GNKurtlar** [reply]
It looks like this doesn't work for Reveal.js slides

**Reply by peekxc** [reply]
Yeah this is related to an ever growing issue in that it's extremely ambiguous what works in quarto.yml vs. the frontmatter of .qmd files.
